src/analysis/souffle/check_predicate.dl

```diff
@@ -0,0 +1,82 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#ifndef SRC_ANALYSIS_SOUFFLE_CHECK_PREDICATE_DL_
+#define SRC_ANALYSIS_SOUFFLE_CHECK_PREDICATE_DL_
+
+#include "authorization_logic.dl"
+#include "dataflow_graph.dl"
+#include "tags.dl"
+
+.type Bool <: number
+
+.type CheckPredicate =
+  CP_TagPresence { ap: AccessPath, principal: Principal, tag: Tag }
+  | CP_And { lhs: CheckPredicate, rhs: CheckPredicate }
+  | CP_Or { lhs: CheckPredicate, rhs: CheckPredicate }
+  | CP_Implies { antecedent: CheckPredicate, consequent: CheckPredicate }
+  | CP_Not { to_negate: CheckPredicate }
```

Comment:
Consider `CP_Not { some_predicate }`. If `some_predicate` is `true`, will  `CP_Not { some_predicate }` be true or false?

Reply:
The result of `CP_Not` will be `false` if its input is `true`.

Reply:
Can you document it? It is not clear from the `to_negate` name.

Reply:
Done.